api/src/main/java/org/apache/iceberg/transforms/Timestamps.java

```diff
@@ -27,48 +27,72 @@
 import org.apache.iceberg.expressions.Expression;
 import org.apache.iceberg.expressions.Expressions;
 import org.apache.iceberg.expressions.UnboundPredicate;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 import org.apache.iceberg.types.Type;
 import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.SerializableFunction;
 
 enum Timestamps implements Transform<Long, Integer> {
   YEAR(ChronoUnit.YEARS, "year"),
   MONTH(ChronoUnit.MONTHS, "month"),
   DAY(ChronoUnit.DAYS, "day"),
   HOUR(ChronoUnit.HOURS, "hour");
 
+  static class Apply implements SerializableFunction<Long, Integer> {
+    private final ChronoUnit granularity;
+
+    Apply(ChronoUnit granularity) {
+      this.granularity = granularity;
+    }
+
+    @Override
+    public Integer apply(Long timestampMicros) {
+      if (timestampMicros == null) {
+        return null;
+      }
+
+      if (timestampMicros >= 0) {
+        OffsetDateTime timestamp =
+            Instant.ofEpochSecond(
```

Comment:
Optional: I know it was copied from another place but it is a bit hard to read this block because of formatting, I'd consider adding temp vars.

```
if (timestampMicros >= 0) {
  long epochSecond = Math.floorDiv(timestampMicros, 1_000_000);
  int nanoAdjustment = Math.floorMod(timestampMicros, 1_000_000) * 1000;
  Instant instant = Instant.ofEpochSecond(epochSecond, nanoAdjustment);
  return (int) granularity.between(EPOCH, instant.atOffset(ZoneOffset.UTC));
} else {
  // ...
  long epochSecond = Math.floorDiv(timestampMicros, 1_000_000);
  int nanoAdjustment = Math.floorMod(timestampMicros + 1, 1_000_000) * 1000;
  Instant instant = Instant.ofEpochSecond(epochSecond, nanoAdjustment);
  return (int) granularity.between(EPOCH, instant.atOffset(ZoneOffset.UTC);) - 1;
}
```

Reply:
I decided not to modify this. I think it's a good suggestion, but messing this up with a cut & paste error just would not be worth it :sweat_smile:.